DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp

```diff
@@ -0,0 +1,409 @@
+  };
+
+  std::map<std::array<int, 3>, std::vector<ArrowFragment>> m_columns;
+};
+
+void registerArrowCsvForeignStorage(void) {
+  ForeignStorageInterface::registerPersistentStorageInterface(
+      new ArrowCsvForeignStorage());
+}
+
+ArrowCsvForeignStorage::~ArrowCsvForeignStorage() {
+  m_columns.clear();
+}
+
+void ArrowCsvForeignStorage::append(
+    const std::vector<ForeignStorageColumnBuffer>& column_buffers) {
+  CHECK(false);
+}
+
+void ArrowCsvForeignStorage::read(const ChunkKey& chunk_key,
+                                  const SQLTypeInfo& sql_type,
+                                  int8_t* dest,
+                                  const size_t numBytes) {
+  std::array<int, 3> col_key{chunk_key[0], chunk_key[1], chunk_key[2]};
+  auto& frag = m_columns.at(col_key).at(chunk_key[3]);
+
+  CHECK(!frag.chunks.empty() || !chunk_key[3]);
+  int64_t sz = 0, copied = 0;
+  arrow::ArrayData* prev_data = nullptr;
+  int varlen_offset = 0;
+
+  for (auto array_data : frag.chunks) {
+    arrow::Buffer* bp = nullptr;
+    if (sql_type.is_dict_encoded_string()) {
+      // array_data->buffers[1] stores dictionary indexes
+      bp = array_data->buffers[1].get();
+    } else if (sql_type.get_type() == kTEXT) {
+      CHECK_GE(array_data->buffers.size(), 3UL);
+      // array_data->buffers[2] stores string array
+      bp = array_data->buffers[2].get();
+    } else if (array_data->null_count != array_data->length) {
+      // any type except strings (none encoded strings offsets go here as well)
+      CHECK_GE(array_data->buffers.size(), 2UL);
+      bp = array_data->buffers[1].get();
+    }
+    if (bp) {
+      // offset buffer for none encoded strings need to be merged as arrow chunkes sizes
+      // are less then omnisci fragments sizes
+      if (chunk_key.size() == 5 && chunk_key[4] == 2) {
+        auto data = reinterpret_cast<const uint32_t*>(bp->data());
+        auto dest_ui32 = reinterpret_cast<uint32_t*>(dest);
+        sz = bp->size();
+        // We merge arrow chunks with string offsets into a single contigous fragment.
+        // Each string is represented by a pair of offsets, thus size of offset table is
+        // num strings + 1. When merging two chunks, the last number in the first chunk
+        // duplicates the first number in the second chunk, so we skip it.
+        if (prev_data && sz > 0) {
+          data++;
+          sz -= sizeof(uint32_t);
+        }
+        if (sz > 0) {
+          // We also re-calculate offsets in the second chunk as it is a continuation of
+          // the first one.
+          std::transform(data,
+                         data + (sz / sizeof(uint32_t)),
+                         dest_ui32,
+                         [varlen_offset](uint32_t val) { return val + varlen_offset; });
+          varlen_offset += data[(sz / sizeof(uint32_t)) - 1];
+        }
+      } else {
+        std::memcpy(dest, bp->data(), sz = bp->size());
+      }
+    } else {
+      // TODO: nullify?
```

Comment:
```suggestion
      // should the dest memory be zeroed? It looks like it is already filled with zeros. This will be irrelevant when copy is eliminated
```